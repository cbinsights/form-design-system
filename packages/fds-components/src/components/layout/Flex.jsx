import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

/**
 * @description map of alignment prop values to class names
 */
const DirectionPropMap = {
  row: 'flex--row',
  column: 'flex--column',
};

/**
 * @description map of justify prop values to class names
 */
const JustifyPropMap = {
  end: 'flex--justifyEnd',
  center: 'flex--justifyCenter',
  spaceBetween: 'flex--justifySpaceBetween',
  spaceAround: 'flex--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
const AlignPropMap = {
  start: 'flex--alignStart',
  end: 'flex--alignEnd',
  center: 'flex--alignCenter',
  stretch: 'flex--alignStretch',
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const Flex = (props) => {
  const classNames = cx(
    'flex',
    AlignPropMap[props.align],
    DirectionPropMap[props.direction],
    {
      [`flex--${props.direction}--reverse`]: props.reverse,
    },
    {
      [JustifyPropMap[props.justify]]: props.justify,
    },
    {
      'flex--wrap': props.wrap,
      'flex--noGutters': props.noGutters,
    },
    props.className
  );

  return <div className={classNames}>{props.children}</div>;
};

Flex.defaultProps = {
  direction: 'row',
  align: 'stretch',
};

Flex.propTypes = {
  direction: PropTypes.oneOf(Object.keys(DirectionPropMap)),
  justify: PropTypes.oneOf(Object.keys(JustifyPropMap)),
  align: PropTypes.oneOf(Object.keys(AlignPropMap)),

  /** When set, gutters are removed from `FlexItem` children */
  noGutters: PropTypes.bool,

  /** When set, `FlexItem` children are allowed to wrap */
  wrap: PropTypes.bool,

  /** When set, `FlexItem` order is rendered in reverse */
  reverse: PropTypes.bool,

  /** Classes to pass to flex parent */
  className: PropTypes.string,

  /** React children (should be of type `FlexItem`) */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
    .isRequired,
};

export default Flex;
