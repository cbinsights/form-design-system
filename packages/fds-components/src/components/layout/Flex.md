
### Flex direction
`Flex` renders in row orientation by default. Use the `direction` prop to change flex direction. To reverse the row or column, use the boolean prop `reverse`.

```js
<div style={{ marginTop: '1rem' }}>
  <p><strong>row</strong></p>
  <Flex direction="row" className="debug--boxModel">
    <FlexItem>
      <img src="https://place-hold.it/48x24/a02385/fff&text=item1" />
    </FlexItem>
    <FlexItem>
      <img src="https://place-hold.it/48x24/a02385/fff&text=item2" />
    </FlexItem>
    <FlexItem>
      <img src="https://place-hold.it/48x24/a02385/fff&text=item3" />
    </FlexItem>
  </Flex>
</div>

<div style={{ marginTop: '1rem' }}>
  <p><strong>row</strong> with <strong>reverse</strong></p>
  <Flex direction="row" reverse className="debug--boxModel">
    <FlexItem>
      <img src="https://place-hold.it/48x24/a02385/fff&text=item1" />
    </FlexItem>
    <FlexItem>
      <img src="https://place-hold.it/48x24/a02385/fff&text=item2" />
    </FlexItem>
    <FlexItem>
      <img src="https://place-hold.it/48x24/a02385/fff&text=item3" />
    </FlexItem>
  </Flex>
</div>

<div style={{ height: '400px', marginTop: '1rem' }}>
  <p><strong>column</strong></p>
  <Flex direction="column" className="debug--boxModel">
    <FlexItem>
      <img src="https://place-hold.it/48x24/a02385/fff&text=item1" />
    </FlexItem>
    <FlexItem>
      <img src="https://place-hold.it/48x24/a02385/fff&text=item2" />
    </FlexItem>
    <FlexItem>
      <img src="https://place-hold.it/48x24/a02385/fff&text=item3" />
    </FlexItem>
  </Flex>
</div>
```

### Flex alignment/justification
Following the flexbox spec, `align` controls cross-axis alignment and `justify` controls main axis alignment. In this example, both `align` and `justify` are set to `end`. The child
`FlexItem` components here are set to `shrink` to their content width.

_See `FlexItem` documentation below for more information on the `shrink` prop_

```js
<Flex align="end" justify="end" className="debug--boxModel">
  <FlexItem shrink>
    <img src="https://place-hold.it/48x24/a02385/fff&text=item1" />
  </FlexItem>
  <FlexItem shrink>
    <img src="https://place-hold.it/48x64/a02385/fff&text=item2" />
  </FlexItem>
  <FlexItem shrink>
    <img src="https://place-hold.it/48x24/a02385/fff&text=item3" />
  </FlexItem>
  <FlexItem shrink>
    <img src="https://place-hold.it/48x64/a02385/fff&text=item4" />
  </FlexItem>
</Flex>
```
