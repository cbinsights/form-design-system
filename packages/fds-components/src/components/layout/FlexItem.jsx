import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

/**
 * @description map of justify prop values to class names
 */
const JustifyPropMap = {
  start: 'flexItem--justifyStart',
  end: 'flexItem--justifyEnd',
  center: 'flexItem--justifyCenter',
  spaceBetween: 'flexItem--justifySpaceBetween',
  spaceAround: 'flexItem--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
const AlignPropMap = {
  start: 'flexItem--alignStart',
  end: 'flexItem--alignEnd',
  center: 'flexItem--alignCenter',
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const FlexItem = (props) => {
  const classNames = cx(
    'flexItem',
    {
      'flexItem--shrink': props.shrink,
    },
    {
      [JustifyPropMap[props.justify]]: props.justify,
      [AlignPropMap[props.align]]: props.align,
    },
    props.className
  );

  return <div className={classNames}>{props.children}</div>;
};

FlexItem.defaultProps = {
  shrink: false,
};

FlexItem.propTypes = {
  justify: PropTypes.oneOf(Object.keys(JustifyPropMap)),
  align: PropTypes.oneOf(Object.keys(AlignPropMap)),

  /** When set, the `FlexItem` will shrink to content size */
  shrink: PropTypes.bool,

  /** Classes to pass to flex child */
  className: PropTypes.string,

  /** React children */
  children: PropTypes.node,
};

export default FlexItem;
