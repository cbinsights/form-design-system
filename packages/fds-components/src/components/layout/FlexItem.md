
### Shrinking & growing a `FlexItem`
By default, all `FlexItem` components will grow to fill remaining space within the parent
`Flex`. The boolean prop `shrink` will cause a `FlexItem` to shrink to its content size.

```js
<Flex align="center" className="debug--boxModel">
  <FlexItem shrink>
    <img src="https://place-hold.it/60x32/a02385/fff&text=shrink" />
  </FlexItem>
  <FlexItem>
    <img src="https://place-hold.it/60x32/a02385/fff&text=default" />
  </FlexItem>
  <FlexItem>
    <img src="https://place-hold.it/60x32/a02385/fff&text=default" />
  </FlexItem>
  <FlexItem shrink>
    <img src="https://place-hold.it/60x32/a02385/fff&text=shrink" />
  </FlexItem>
</Flex>
```

### Controlling alignment/justification
You can use the same `align` and `justify` prop interface from `Flex` to control alignment and justification of individual `FlexItem` components.

```js
<Flex className="debug--boxModel">
  <FlexItem shrink>
    <img src="https://place-hold.it/60x80/a02385/fff&text=default" />
  </FlexItem>
  <FlexItem shrink align="end">
    <img src="https://place-hold.it/60x32/a02385/fff&text=end" />
  </FlexItem>
  <FlexItem shrink align="start">
    <img src="https://place-hold.it/60x32/a02385/fff&text=start" />
  </FlexItem>
  <FlexItem shrink>
    <img src="https://place-hold.it/60x80/a02385/fff&text=default" />
  </FlexItem>
</Flex>
```

## Flex & FlexItem cookbook

### Auto-sized content area with fixed footer
Within a fixed height container, this `Flex` creates a fixed height header and footer and allows the content area to expand to fill the remaining space.

```js
<div style={{ height: '200px' }}>
  <Flex direction="column" align="stretch" className="debug--boxModel">
    <FlexItem shrink>
      <p>This is the header, set to <code>shrink</code>.</p>
    </FlexItem>
    <FlexItem>
      <p>This is the content area. It grows to fill remaining space regardless of how much content this <code>FlexItem</code> contains.</p>
    </FlexItem>
    <FlexItem shrink>
      <p>This is the footer, set to <code>shrink</code>.</p>
    </FlexItem>
  </Flex>
</div>
```

### Comment block
Basic layout of comments in the CB Insights platform.

```js
<Flex>
  <FlexItem shrink>
    <img src="https://place-hold.it/48x48/a02385/fff" />
  </FlexItem>
  <FlexItem>
    <div style={{ 
      border: '1px solid #d6d4d4', 
      padding: '0.5rem', 
      borderRadius: '6px' 
    }}>
      <p><strong>A dog</strong></p>
      <p>bark bark bork buff boark bork bark berk woof</p>
      <p>This content expands to fill remaining space within the <code>Flex</code></p>
    </div>
    <p><a href="#">Reply</a> &middot; <a href="#">Delete</a></p>
  </FlexItem>
</Flex>
```

### Flexible site header
`Flex` and `FlexItem` provide advantages in flexibility; fixed widths and explicit padding/marging are less necessary, and it's easy to add and remove elements to a row without the layout breaking.

_Open the editor below to try adding and removing items_

```js
<Flex align="center">
  <FlexItem shrink>
    <img src="https://place-hold.it/36x36/a02385/fff?text=menu" />
  </FlexItem>
  <FlexItem shrink>
    <img src="https://place-hold.it/200x36/a02385/fff?text=logo" />
  </FlexItem>
  <FlexItem>
    <input
      type="text"
      placeholder="Search"
      style={{ width: '100%', lineHeight: '30px' }}
    />
  </FlexItem>
  <FlexItem shrink>
    <img src="https://place-hold.it/36x36/a02385/fff?text=icon" />
  </FlexItem>
  <FlexItem shrink>
    <img src="https://place-hold.it/36x36/a02385/fff?text=icon" />
  </FlexItem>
</Flex>
```

### FLEXIBLE üëè LAYOUT üëè CHALLENGE
üö® LIVE CODING DEMO üö®

```js
```
