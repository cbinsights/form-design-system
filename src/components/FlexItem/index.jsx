import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

/**
 * @description map of justify prop values to class names
 */
const JustifyPropMap = {
  start: 'flexItem--justifyStart',
  end: 'flexItem--justifyEnd',
  center: 'flexItem--justifyCenter',
  spaceBetween: 'flexItem--justifySpaceBetween',
  spaceAround: 'flexItem--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
const AlignPropMap = {
  start: 'flexItem--alignStart',
  end: 'flexItem--alignEnd',
  center: 'flexItem--alignCenter',
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const FlexItem = ({ shrink = false, justify, align, children }) => {
  const classNames = cx(
    'flexItem',
    {
      'flexItem--shrink': shrink,
    },
    {
      [JustifyPropMap[justify]]: justify,
      [AlignPropMap[align]]: align,
    }
  );

  return <div className={classNames}>{children}</div>;
};

FlexItem.propTypes = {
  /** Sets standard justify-content */
  justify: PropTypes.oneOf(Object.keys(JustifyPropMap)),

  /** Sets standard align-items */
  align: PropTypes.oneOf(Object.keys(AlignPropMap)),

  /** When set, the `FlexItem` will shrink to content size */
  shrink: PropTypes.bool,

  /** React children */
  children: PropTypes.node,
};

export default FlexItem;
