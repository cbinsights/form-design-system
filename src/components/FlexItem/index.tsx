import React from 'react';
import cc from 'classcat';

/**
 * @description map of justify prop values to class names
 */
const JustifyPropMap = {
  start: 'flexItem--justifyStart',
  end: 'flexItem--justifyEnd',
  center: 'flexItem--justifyCenter',
  spaceBetween: 'flexItem--justifySpaceBetween',
  spaceAround: 'flexItem--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
const AlignPropMap = {
  start: 'flexItem--alignStart',
  end: 'flexItem--alignEnd',
  center: 'flexItem--alignCenter',
};

export interface FlexItemProps {
  /** Sets standard justify-content */
  justify?: 'start' | 'end' | 'center' | 'spaceBetween' | 'spaceAround';

  /** Sets standard align-items */
  align?: 'start' | 'end' | 'center';

  /** When set, the `FlexItem` will shrink to content size */
  shrink?: boolean;

  /** React children */
  children?: React.ReactNode;
}

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const FlexItem = ({
  shrink = false,
  justify,
  align,
  children,
}: FlexItemProps): JSX.Element => {
  const classNames = cc([
    align && AlignPropMap[align],
    justify && JustifyPropMap[justify],
    {
      'flexItem--shrink': shrink,
    },
    'flexItem',
  ]);

  return <div className={classNames}>{children}</div>;
};

export default FlexItem;
