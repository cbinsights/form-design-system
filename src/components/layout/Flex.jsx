import React from 'react';
import cc from 'classcat';
import PropTypes from 'prop-types';

/**
 * @description map of alignment prop values to class names
 */
export const DirectionPropMap = {
  row: 'flex--row',
  column: 'flex--column',
};

/**
 * @description map of justify prop values to class names
 */
export const JustifyPropMap = {
  end: 'flex--justifyEnd',
  center: 'flex--justifyCenter',
  spaceBetween: 'flex--justifySpaceBetween',
  spaceAround: 'flex--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
export const AlignPropMap = {
  start: 'flex--alignStart',
  end: 'flex--alignEnd',
  center: 'flex--alignCenter',
  stretch: 'flex--alignStretch',
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const Flex = ({
  align = 'stretch',
  direction = 'row',
  reverse,
  justify,
  wrap,
  noGutters,
  children,
}) => {
  const classNames = cc([
    {
      [`flex--${direction}--reverse`]: reverse,
    },
    {
      [JustifyPropMap[justify]]: justify,
    },
    {
      'flex--wrap': wrap,
      'flex--noGutters': noGutters,
    },
    'flex',
    AlignPropMap[align],
    DirectionPropMap[direction],
  ]);

  return <div className={classNames}>{children}</div>;
};

Flex.propTypes = {
  /** sets flex-direction (along with either 100% height or width) */
  direction: PropTypes.oneOf(Object.keys(DirectionPropMap)),

  /** sets standard justify-content */
  justify: PropTypes.oneOf(Object.keys(JustifyPropMap)),

  /** sets standard align-items */
  align: PropTypes.oneOf(Object.keys(AlignPropMap)),

  /** When set, gutters are removed from `FlexItem` children */
  noGutters: PropTypes.bool,

  /** When set, `FlexItem` children are allowed to wrap */
  wrap: PropTypes.bool,

  /** When set, `FlexItem` order is rendered in reverse */
  reverse: PropTypes.bool,

  /** React children (should be of type `FlexItem`) */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
    .isRequired,
};

export default Flex;
