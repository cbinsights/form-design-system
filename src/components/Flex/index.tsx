import React from 'react';
import cc from 'classcat';
import { FlexItemProps } from '../FlexItem';

/**
 * @description map of alignment prop values to class names
 */
export const DirectionPropMap = {
  row: 'flex--row',
  column: 'flex--column',
};

/**
 * @description map of justify prop values to class names
 */
export const JustifyPropMap = {
  end: 'flex--justifyEnd',
  center: 'flex--justifyCenter',
  spaceBetween: 'flex--justifySpaceBetween',
  spaceAround: 'flex--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
export const AlignPropMap = {
  start: 'flex--alignStart',
  end: 'flex--alignEnd',
  center: 'flex--alignCenter',
  stretch: 'flex--alignStretch',
};

export interface FlexProps {
  /** sets flex-direction (along with either 100% height or width) */
  direction?: 'row' | 'column';

  /** switches flex-direction at a given breakpoint */
  switchDirection?: 'xs' | 's' | 'm' | 'l' | 'xl';

  /** sets standard justify-content */
  justify?: 'end' | 'center' | 'spaceBetween' | 'spaceAround';

  /** sets standard align-items */
  align?: 'start' | 'end' | 'center' | 'stretch';

  /** When set, gutters are removed from `FlexItem` children */
  noGutters?: boolean;

  /** When set, `FlexItem` order is rendered in reverse */
  reverse?: boolean;

  /** React children (should be of type `FlexItem`) */
  children: React.ReactElement<FlexItemProps> | Array<React.ReactElement<FlexItemProps>>;
}

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const Flex = ({
  align = 'stretch',
  direction = 'row',
  switchDirection: breakpoint,
  reverse,
  justify,
  noGutters,
  children,
}: FlexProps): JSX.Element => {
  const classNames = cc([
    'flex',
    AlignPropMap[align],
    DirectionPropMap[direction],
    justify && JustifyPropMap[justify],
    {
      [`${breakpoint}:flex--${direction === 'row' ? 'column' : 'row'}`]: breakpoint,
    },
    {
      'flex--reverse': reverse,
    },
    {
      'flex--noGutters': noGutters,
    },
  ]);

  return <div className={classNames}>{children}</div>;
};

export default Flex;
