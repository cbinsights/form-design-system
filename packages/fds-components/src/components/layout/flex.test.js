import React from 'react';
import { shallow } from 'enzyme';

import Flex from './Flex';
import FlexItem from './FlexItem';

const renderComponent = (props) => shallow(<Flex {...props}><FlexItem /></Flex>);

describe('Flex component', () => {

  it('matches snapshot (default props)', () => {
    const flex = renderComponent();
    expect(flex).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const flex = renderComponent({
      noGutters: true,
      wrap: true,
      reverse: true,
      direction: 'column',
      align: 'end',
      justify: 'spaceBetween',
      className: 'foo'
    });
    expect(flex).toMatchSnapshot();
  });
});
