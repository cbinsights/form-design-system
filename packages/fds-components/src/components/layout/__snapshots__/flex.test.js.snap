// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex component matches snapshot (default props) 1`] = `
<div
  className="flex flex--alignStretch flex--row"
>
  <FlexItem
    shrink={false}
  />
</div>
`;

exports[`Flex component matches snapshot (set all props) 1`] = `
<div
  className="flex flex--alignEnd flex--column flex--column--reverse flex--justifySpaceBetween flex--wrap flex--noGutters foo"
>
  <FlexItem
    shrink={false}
  />
</div>
`;
